presto-common/src/main/java/com/facebook/presto/common/type/SqlTimestamp.java

```diff
@@ -15,47 +15,78 @@
 
 import com.fasterxml.jackson.annotation.JsonValue;
 
-import java.time.Instant;
 import java.time.ZoneId;
 import java.time.format.DateTimeFormatter;
 import java.util.Objects;
 import java.util.Optional;
 
 import static com.facebook.presto.common.type.TimeZoneKey.UTC_KEY;
+import static com.facebook.presto.common.type.TimestampMicrosUtils.microsToInstant;
+import static java.lang.Math.floorDiv;
 
 public final class SqlTimestamp
 {
     // This needs to be Locale-independent, Java Time's DateTimeFormatter compatible and should never change, as it defines the external API data format.
-    public static final String JSON_FORMAT = "uuuu-MM-dd HH:mm:ss.SSS";
+    public static final String JSON_FORMAT = "uuuu-MM-dd HH:mm:ss.SSSSSS";
     public static final DateTimeFormatter JSON_FORMATTER = DateTimeFormatter.ofPattern(JSON_FORMAT);
 
-    private final long millis;
+    private final long micros;
     private final Optional<TimeZoneKey> sessionTimeZoneKey;
 
+    public static SqlTimestamp ofMicros(long micros)
+    {
+        return new SqlTimestamp(micros, Optional.empty());
+    }
+
+    public static SqlTimestamp ofMicrosLegacy(long micros, TimeZoneKey timeZoneKey)
+    {
+        return new SqlTimestamp(micros, Optional.of(timeZoneKey));
+    }
+
+    @Deprecated
     public SqlTimestamp(long millis)
     {
-        this.millis = millis;
+        this.micros = millis * 1000;
         sessionTimeZoneKey = Optional.empty();
     }
 
     @Deprecated
     public SqlTimestamp(long millisUtc, TimeZoneKey sessionTimeZoneKey)
     {
-        this.millis = millisUtc;
+        this.micros = millisUtc * 1000;
         this.sessionTimeZoneKey = Optional.of(sessionTimeZoneKey);
     }
 
+    private SqlTimestamp(long micros, Optional<TimeZoneKey> sessionTimeZoneKey)
+    {
+        this.micros = micros;
+        this.sessionTimeZoneKey = sessionTimeZoneKey;
```

Comment:
nit: perhaps, add requireNotNull(sessionTimeZoneKey, "sessionTimeZoneKey is null")